packages/sdk/src/protocols/base_protocol.ts

```diff
@@ -1,15 +1,206 @@
+import type { Peer, PeerId } from "@libp2p/interface";
+import { ConnectionManager } from "@waku/core";
+import { BaseProtocol } from "@waku/core/lib/base_protocol";
 import { IBaseProtocolSDK } from "@waku/interfaces";
+import { Logger } from "@waku/utils";
 
 interface Options {
   numPeersToUse?: number;
+  maintainPeersInterval?: number;
 }
 
 const DEFAULT_NUM_PEERS_TO_USE = 3;
+const DEFAULT_MAINTAIN_PEERS_INTERVAL = 30_000;
 
 export class BaseProtocolSDK implements IBaseProtocolSDK {
-  public readonly numPeers: number;
+  public readonly numPeersToUse: number;
+  private peers: Peer[] = [];
+  private maintainPeersIntervalId: NodeJS.Timeout | undefined;
+  log = new Logger("sdk:base-protocol");
 
-  constructor(options: Options) {
-    this.numPeers = options?.numPeersToUse ?? DEFAULT_NUM_PEERS_TO_USE;
+  maintainPeersLock = false;
+
+  constructor(
+    protected core: BaseProtocol,
+    private connectionManager: ConnectionManager,
+    options: Options
+  ) {
+    this.log = new Logger(`sdk:${core.multicodec}`);
+    this.numPeersToUse = options?.numPeersToUse ?? DEFAULT_NUM_PEERS_TO_USE;
+    const maintainPeersInterval =
+      options?.maintainPeersInterval ?? DEFAULT_MAINTAIN_PEERS_INTERVAL;
+
+    this.startMaintainPeersInterval(maintainPeersInterval).catch((error) => {
+      this.log.error("Error starting maintain peers interval:", error);
+      throw error;
+    });
+  }
+
+  get connectedPeers(): Peer[] {
+    return this.peers;
+  }
+
+  /**
+   * Disconnects from a peer and tries to find a new one to replace it.
+   * @param peerToDisconnect The peer to disconnect from.
+   */
+  public async renewPeer(peerToDisconnect: PeerId): Promise<void> {
+    this.log.info(`Renewing peer ${peerToDisconnect}`);
+    try {
+      await this.connectionManager.dropConnection(peerToDisconnect);
+      this.peers = this.peers.filter((peer) => peer.id !== peerToDisconnect);
```

Comment:
I am curious if we need to support a copy of `peers` in general, as if possible I would try to rely on `peerStore` or some single source we use for such cases

Reply:
why can't we mark them as used in the `PeerStore`? - this way we have only one source of truth

Reply:
Good point about `PeerStore` not having any info about connections. I think in terms of `libp2p` it seems like a phone book - no sense of bringing other info as it gets harder to maintain over time

Let's use local copy for now then, we can optimize it later.